src/Parallel/Reduction.hpp

```diff
@@ -9,36 +9,87 @@
+#include "Utilities/Functional.hpp"
+#include "Utilities/Gsl.hpp"
 #include "Utilities/Requires.hpp"
 #include "Utilities/TypeTraits.hpp"
 
 namespace Parallel {
+/// \cond
+template <class... Ts>
+struct ReductionData;
+/// \endcond
+
+namespace detail {
 /*!
  * \ingroup ParallelGroup
- * \brief Used for reducing heterogeneous collection of types in a single
- * reduction call
+ * \brief Convert a `ReductionData` to a `CkReductionMsg`. Used in custom
+ * reducers.
+ */
+template <class... Ts>
+CkReductionMsg* new_reduction_msg(
+    ReductionData<Ts...>& reduction_data) noexcept {
+  return CkReductionMsg::buildNew(static_cast<int>(reduction_data.size()),
+                                  reduction_data.packed().get());
+}
+}  // namespace detail
+
+/*!
+ * \ingroup ParallelGroup
+ * \brief The data to be reduced, and invokables to be called whenever two
+ * reduction messages are combined and after the reduction has been completed.
  *
- * For each `ReductionData` you must write a custom reducer to reduce the data.
- * Here is an example of such a function:
+ * `InvokeCombine` is a binary invokable that maps `(T current_state, T element)
+ * -> T`, where the `current_state` is the result of reductions so far. The
+ * `InvokeFinal` is an n-ary that takes as its first argument a `T
+ * result_of_reduction` and is invoked once after the reduction is completed.
+ * The additional arguments correspond to the resultant data of earlier
+ * ReductionDatum template parameters in the ReductionData, and are identified
+ * via the `InvokeFinalExtraArgsIndices`, which must be a `std::index_sequence`.
+ * For example, consider computing the L2 error of the evolved variables `u` and
+ * `v`. The ReductionData would look like:
  *
- * \snippet Test_AlgorithmReduction.cpp custom_reduce_function
+ * \code
+ * ReductionData<ReductionDatum<size_t, std::plus<>, Identity<>>,
+ *               ReductionDatum<double, SquaredPlus, SqrtDiv, // for u
+ *                              std::index_sequence<0>>,
+ *               ReductionDatum<double, SquaredPlus, SqrtDiv, // for v
+ *                              std::index_sequence<0>>>
+ *      reduction_data{num_pts, u_err, v_err};
+ * \endcode
  *
- * A variable number of `CkReductionMsg*`s are passed to the reducer, and so it
- * is necessary to loop over each message, contributing the data into a new
- * `ReductionData` (called `reduced` in the above example). The
- * `Parallel::get()` function is used to retrieve elements from the
- * `ReductionData` in a manner that is completely analogous to retrieving
- * elements from a `std::tuple` using `std::get`. Once the new `ReductionData`
- * has been filled, a new `CkReductionMsg*` must be returned using
- * `Parallel::new_reduction_msg()`.
+ * where `SquaredPlus` does `return state += square(element);`, and `SqrtDiv`
+ * does `(error_squared, total_num_pts) { return sqrt(error_squared) /
+ * total_num_pts; }`, `total_num_pts` is the total number of points after the
```

Comment:
It sounds like these SquaredPlus and SqrtDiv exist, but I can’t find them so they are expected to be user-provided? Again, I don’t quite see the benefit of writing those when I can just do the computation in the receiving action.

Reply:
See my response to the other time you asked this. If there's interest I can add these to `Functional.hpp`